esp-hal-common/src/gpio.rs

```diff
@@ -1642,3 +1646,119 @@ pub(crate) use gpio;
+
+    use super::*;
+    use crate::prelude::*;
+
+    #[allow(clippy::declare_interior_mutable_const)]
+    const NEW_AW: AtomicWaker = AtomicWaker::new();
+    static PIN_WAKERS: [AtomicWaker; NUM_PINS] = [NEW_AW; NUM_PINS];
+
+    impl<MODE, RA, IRA, PINTYPE, SIG, const GPIONUM: u8> Wait
+        for GpioPin<Input<MODE>, RA, IRA, PINTYPE, SIG, GPIONUM>
+    where
+        RA: BankGpioRegisterAccess,
+        PINTYPE: IsOutputPin,
+        IRA: InteruptStatusRegisterAccess,
+        SIG: GpioSignal,
+    {
+        async fn wait_for_high(&mut self) -> Result<(), Self::Error> {
+            PinFuture::new(self, Event::HighLevel).await
+        }
+
+        async fn wait_for_low(&mut self) -> Result<(), Self::Error> {
+            PinFuture::new(self, Event::LowLevel).await
+        }
+
+        async fn wait_for_rising_edge(&mut self) -> Result<(), Self::Error> {
+            PinFuture::new(self, Event::RisingEdge).await
+        }
+
+        async fn wait_for_falling_edge(&mut self) -> Result<(), Self::Error> {
+            PinFuture::new(self, Event::FallingEdge).await
+        }
+
+        async fn wait_for_any_edge(&mut self) -> Result<(), Self::Error> {
+            PinFuture::new(self, Event::AnyEdge).await
+        }
+    }
+
+    pub struct PinFuture<'a, P> {
+        pin: &'a mut P,
+    }
+
+    impl<'a, P> PinFuture<'a, P>
+    where
+        P: crate::gpio::Pin + embedded_hal_1::digital::ErrorType,
+    {
+        pub fn new(pin: &'a mut P, event: Event) -> Self {
+            pin.listen(event);
+            Self { pin }
+        }
+    }
+
+    impl<'a, P> core::future::Future for PinFuture<'a, P>
+    where
+        P: crate::gpio::Pin + embedded_hal_1::digital::ErrorType,
+    {
+        type Output = Result<(), P::Error>;
+
+        fn poll(self: core::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            PIN_WAKERS[self.pin.number() as usize].register(cx.waker());
+
+            // if pin is no longer listening its been triggered
+            // therefore the future has resolved
+            if !self.pin.is_listening() {
+                Poll::Ready(Ok(()))
+            } else {
+                Poll::Pending
+            }
+        }
+    }
+
+    #[interrupt]
+    unsafe fn GPIO() {
+        // TODO how to handle dual core reg access
+        // we need to check which core the interrupt is currently firing on
+        // and only fire interrupts registered for that core
+        type Bank0 = SingleCoreInteruptStatusRegisterAccessBank0;
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        type Bank1 = SingleCoreInteruptStatusRegisterAccessBank1;
+
+        let mut intrs = Bank0::pro_cpu_interrupt_status_read() as u64;
+
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        {
+            intrs |= (Bank1::pro_cpu_interrupt_status_read() as u64) << 32;
+        }
+
+        // clear interrupts
+        Bank0GpioRegisterAccess::write_interrupt_status_clear(!0);
```

Comment:
@MabezDev I'm not entirely sure but I think that `set_int_enable` does not implicitly clear interrupt status bits, it is possible for an edge-triggered listener to wake a task twice (in this call, and at a later GPIO interrupt request).

Reply:
I don't quite understand your first point, but I do agree that we should only clear the interrupts read from the status inside the GPIO interrupt handler.

Reply:
I am in the process of opening an issue with more data that confirms my suspicions: level-triggered interrupts fire multiple times, it looks like.